### PR TITLE
Pitch black mitigation test

### PR DESCRIPTION
**Pull request by ArneBab**
Preliminary Evaluation:

*Without mitigation, the pitch black attack is devastating:*

results-01-a-successful-pitch-black-attack-without-mitigation-disrupts-the-network-node-locations-of-nodes

results-02-a-successful-pitch-black-attack-without-mitigation-disrupts-the-network-peer-locations

results-03-a-successful-pitch-black-attack-without-mitigation-disrupts-routing

*The pitch black mitigation reduces location loss — and therefore data loss — during an ongoing attack:*

results-04-the-mitigation-reduces-location-loss-during-an-ongoing-attack

*After a devastating pitch black attack, the mitigation recovers a reasonable network-structure (the cycles here are more than factor 10 slower than for the other cases):*

results-05-the-mitigation-recovers-a-reasonable-location-distribution-after-a-devastating-pitch-black-attack

results-06-the-mitigation-recovers-routing-efficiency-after-a-devastating-pitch-black-attack

### CODE REVIEW

**Review by Bombe — COMMENTED** [review]
Can we get a README (or an entry in another README) detailing how to use this?

**Comment by ArneBab** [comment]
> Can we get a README (or an entry in another README) detailing how to use this?

I have an article with all the scripts in the works.

Would it suffice (or even be better) to have Javadoc in the test that shows which tweaks could be useful and how to evaluate the results?

File: src/freenet/node/LocationManager.java [hunk 1]
```diff
@@ -189,18 +193,18 @@ public void run() {
...
+                node.ticker.queueTimedJob(this, PITCH_BLACK_MITIGATION_FREQUENCY_ONE_DAY);
                 if (swappingDisabled()) {
                     return;
                 }
-                LocalDateTime now = LocalDateTime.now();
+                LocalDateTime now = LocalDateTime.now(clockForTesting);
                 String isoDateStringToday = DateTimeFormatter.ISO_DATE
                     .format(now);
                 String isoDateStringYesterday = DateTimeFormatter.ISO_DATE
                     .format(now.minus(Duration.ofDays(1)));
```
[the remark this answers is not in the document]
Comment (ArneBab) [reply src/freenet/node/LocationManager.java:204]: Thank you for that catch!

**Comment by Bombe** [comment]
> > Can we get a README (or an entry in another README) detailing how to use this?
> 
> I have an article with all the scripts in the works.
> 
> Would it suffice (or even be better) to have Javadoc in the test that shows which tweaks could be useful and how to evaluate the results?

IMO javadoc is the place for “what this does and how it does it” but I care more about the “how do I use it.”

**Comment by ArneBab** [comment]
now with a note in the simulator readme.txt

**Pull request merged**